### PR TITLE
Processor branch

### PR DESCRIPTION
**Pull request by claireberschauer**
Contains pvpro_post_processor.py,  a module containing a class which takes in the output data frame of PVPRO and defines
methods to process the dataset, perform signal decompositions to model degradation trends, analyze how well the models fit, and visualize the trends.

### CODE REVIEW

File: solardatatools/pvpro_post_processing.py [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+        # is a dictionary
+        # est should only be True when the input dataframe contains 'ref' and 'ref_est'
+        # columns
+        
+        if dates is None:
+            dates = [0]
+        else:
+            dates = dates
+        
+        # attributes
```
Comment (bmeyers) [inline solardatatools/pvpro_post_processing.py:42]: Attributes that are set by __init__ don't need to be "declared" as None earlier in the function. A typical structure for init would be 

- Run setup code snippets
- Set attributes that will be set by __init__
- "Declare" other attributes as None that will be set by other methods

File: solardatatools/pvpro_post_processing.py [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+            self.boundary_to_nan()
+        else:
+            pass
+        self.scale_max_1()
+        self.ln_df()
+        df_p = self.df_l
+        
+        self.df_p = df_p
+    
+    ############################################################################################################
```
Comment (bmeyers) [inline solardatatools/pvpro_post_processing.py:81]: too long

File: solardatatools/pvpro_post_processing.py [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+    def data_setup(self, est=False):
+        # adjusts time index so that there are equal intervals
+        time_delta, count = mode(np.diff(self.df.index))
+        freq = int(time_delta[0] / np.timedelta64(1, 's'))
+        new_index = pd.date_range(start=self.df.index[0], end=self.df.index[-1], freq='{}s'.format(freq))
+        self.df = self.df.reindex(index=new_index)
+        
+        # isolates columns of interest
+        if est is True:
+            cols = [c for c in self.df.columns if 'start' not in c and 'end' not in c and 'mean' not in c and
```
Comment (bmeyers) [inline solardatatools/pvpro_post_processing.py:94]: Logic is backwards. Should be defining based on known keys.

File: solardatatools/pvpro_post_processing.py [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+                self.optimize(label, float(l4), float(l5), model)
+                self.plot_sd_space(label, model, model_title=model_title)
+                
+            if opt == 'n':
+                pass
+        else:        
+            components = self.ScaledData[label + '_' + model]
+            if model_title is None:
+                model_title = model
+            else:
```
Comment (bmeyers) [inline solardatatools/pvpro_post_processing.py:324]: delete lines

File: solardatatools/pvpro_post_processing.py [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+                else:
+                    plt.plot(self.df_p.index, components[col])
+
+                plt.title(titles[counter])
+                plt.xticks(rotation=45)
+
+                counter += 1
+
+            plt.tight_layout()
+            plt.show()
```
Comment (bmeyers) [inline solardatatools/pvpro_post_processing.py:353]: replace this line with `return plt.gcf()`

File: solardatatools/pvpro_post_processing.py [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+    
+    def error_analysis(self, lambda4, lambda5, num_runs, lambda2=0.001, solver='Default'):
+        # calculates the holdout error, looping over system parameters, models, cost function weights,
+        # and number of repetitions
+        # lambda4 and lambda5 inputs are arrays
+        
+        ti = time()
+        period = self.period
+        lambda_2 = lambda2
+        lambda_4_values = lambda4
```
Comment (bmeyers) [inline solardatatools/pvpro_post_processing.py:417]: `np.atleast_1d(lambda4)`

File: solardatatools/pvpro_post_processing.py [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+        lambda_5_values = lambda5
+        runs = np.arange(num_runs)
+        models = ['linear', 'monotonic', 'smooth_monotonic', 'piecewise_linear']
+        cols = ['system_parameter', 'run_number', 'degradation_model', 'lambda_2_val',
```
Comment (bmeyers) [inline solardatatools/pvpro_post_processing.py:421]: this should come from global dict

File: solardatatools/pvpro_post_processing.py [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+                for m_type in models:
+                    for l4_val in lambda_4_values:
+                        if m_type == 'smooth_monotonic' or m_type == 'piecewise_linear':
+                            l5_iter = lambda_5_values   
+                            
+                        else:
+                            l5_iter = [1]
+                            
+                        for l5_val in l5_iter:
+                            if solver is 'Default':
```
Comment (bmeyers) [inline solardatatools/pvpro_post_processing.py:454]: duplicated lines, not necessary

File: solardatatools/pvpro_post_processing.py [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+                            df_error.loc[counter] = row
+                            counter += 1
+
+                            t_progress = time()
+                            msg = "{:.2f} minutes".format((t_progress-ti)/60)
+                            progress(counter, num_rows, status=msg)
+                            
+        self.df_error = df_error
+        
+        ti = time()
```
Comment (bmeyers) [inline solardatatools/pvpro_post_processing.py:476]: please add a comment

File: solardatatools/pvpro_post_processing.py [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+        cols = ['system_parameter', 'degradation_model', 'lambda_2_val', 'lambda_4_val', 
+                'lambda_5_val', 'mean_sq_error']
+        num_rows_avg = len(lambda_4_values)*(len(lambda_5_values) + 1)*2*len(self.df_p.columns)
+        df_error_avg = pd.DataFrame(columns=cols, index=np.arange(num_rows_avg))
+        counter = 0
+        progress(counter, num_rows_avg)
+        
+        for label, values in self.df_p.items():
```
Comment (bmeyers) [inline solardatatools/pvpro_post_processing.py:484]: pandas groupby https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html

**Review by bmeyers — CHANGES_REQUESTED** [review]

**Comment by bmeyers** [comment]
@Thistleman If you get a chance, could you check on why tests are failing on @claireberschauer's branch?

**Comment by bmeyers** [comment]
Thanks for continuing to clean things up, @claireberschauer! Do you think this is ready for final review and merge?

**Comment by claireberschauer** [comment]
> 

I think so! Let me know if you want me to make any more edits or if I missed something. Class doesn't start until Tuesday so I have a bit of free time for the next few days.

**Review by bmeyers — APPROVED** [review]

**Comment by bmeyers** [comment]
This looks great! I'm going to merge it now. I just sent an email about other documents that I'd like to track down before any more time passes. Please let me know if there are any issues.

**Pull request merged**